### PR TITLE
[REVIEW] Fix fillna behaviour with different dtypes

### LINKED ISSUE #1261: [BUG] Series.fillna changes dtype in 0.6
After building `branch-0.6` (3a8dc23) locally I'm seeing the following behavior when using `Series.fillna`:
```python
In [1]: import cudf
In [2]: import numpy as np
In [3]: df = cudf.DataFrame()
In [4]: df['a'] = [1, 2, 3]
In [5]: df.a.astype(np.int16).fillna(1).dtype
Out[5]: dtype('int64')
```
When running same code in latest rapids container (`cudf==0.5`) `Out[5]` above is `dtype('int16')` as expected.

### PR DESCRIPTION
**Pull request by shwina**
Fixes #1261.

### CODE REVIEW

**Comment by shwina** [comment]
This PR helps us match Pandas' `fillna()` behaviour more closely.

Currently, cudf automatically "upcasts" when filling values with a different dtype:

```
In [1]: import cudf                                                                                                                                                                                       

In [2]: a = cudf.Series([1, 2, None, 3])                                                                                                                                                                  

In [3]: print(a.fillna(3.5))                                                                                                                                                                              
0    1.0
1    2.0
2    3.5
3    3.0
dtype: float64
```

After (if) this PR is merged, we will instead raise an error (same as Pandas):

```
In [1]: import cudf                                                                                                                                                                                       

In [2]: a = cudf.Series([1, 2, None, 3])                                                                                                                                                                  

In [3]: print(a.fillna(3.5))                                                                                                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-11f07380e850> in <module>
----> 1 print(a.fillna(3.5))

~/projects/cudf/python/cudf/dataframe/series.py in fillna(self, value, method, axis, inplace, limit)
    654             raise NotImplementedError("The axis keyword is not supported")
    655 
--> 656         data = self._column.fillna(value, inplace=inplace)
    657 
    658         if not inplace:

~/projects/cudf/python/cudf/dataframe/numerical.py in fillna(self, fill_value, inplace)
    449         fill_value_col = columnops.as_column(fill_value, nan_as_null=False)
    450         if is_integer_dtype(result.dtype):
--> 451             fill_value_col = safe_cast_to_int(fill_value_col, result.dtype)
    452         else:
    453             fill_value_col = fill_value_col.astype(result.dtype)

~/projects/cudf/python/cudf/dataframe/numerical.py in safe_cast_to_int(col, dtype)
    511         raise TypeError("Cannot safely cast non-equivalent {} to {}".format(
    512             col.dtype.type.__name__,
--> 513             np.dtype(dtype).type.__name__))
    514 
    515 

TypeError: Cannot safely cast non-equivalent float64 to int64
```

I'm not sure which is more desirable and I would appreciate a second opinion on this.

File: python/cudf/tests/test_replace.py [hunk 1]
```diff
@@ -187,3 +187,10 @@ def test_fillna_dataframe(fill_type, inplace):
         got = gdf
 
     assert_eq(expect, got)
+
+def test_series_fillna_dtype():
+    df = DataFrame()
+    df['a'] = [1, 2, 3]
```
Comment (kkraus14) [inline python/cudf/tests/test_replace.py]: Should probably have a null value in here just to confirm it fills correctly as well 😄

Comment (shwina) [reply python/cudf/tests/test_replace.py]: good idea :P definitely going to be adding a few more tests

**Comment by shwina** [comment]
@kkraus14 This is ready to review but please see my comment above - thanks!

File: python/cudf/tests/test_replace.py [hunk 2]
```diff
@@ -84,23 +86,31 @@ def test_dataframe_replace():
...
     'inplace',
     [True, False])
-def test_series_fillna_numerical(dtype, fill_type, null_value, inplace):
-    data = np.array([0, 1, null_value, 2, null_value], dtype='float64')
-    sr = Series(data).astype(dtype)
+def test_series_fillna_numerical(data_dtype, fill_dtype,
+                                 fill_type, null_value, inplace):
+    # TODO: These tests should use Pandas' nullable int type
+    # when we support a recent enough version of Pandas
+    # https://pandas.pydata.org/pandas-docs/stable/user_guide/integer_na.html
```
Comment (kkraus14) [inline python/cudf/tests/test_replace.py]: We support the most recent version of Pandas AFAIK, but given most users aren't using the special Nullable types yet I don't think this is super high priority.

**Pull request merged**